### ISSUE TITLE
intelligent autocompleter

### ISSUE THREAD

**Issue by pointhi** [searx/searx#191]
It would be greate, if there would be a mode, in which the autcompleter allways using the best choise of autocompletion (as example: `google` for `general`, `photon` for `map`,...)

**Comment by Cqoicebordel** [comment]
I think it would be difficult to implement it in a way it was unobtrusive for the user : I tend to chose the category of the search after I typed my query, so the autocompleter would use the wrong engines to autocomplete my query.

But I like the idea anyway. Maybe we could do it the other way around : in the suggestion list, we could provide icons to indicate to which category this suggestion belong.

**Comment by Glandos** [comment]
> But I like the idea anyway. Maybe we could do it the other way around : in the suggestion list, we could provide icons to indicate to which category this suggestion belong.

I would love it. Really.

**Comment by stef** [comment]
unless you can do an auto-completer without javascript and are ok with extra roundtrip times through the searx server, an auto-completer would greatly harm the prime directive of searx: preserve privacy.

i think by reducing the privacy defaults in searx just to get more users, we do not honor the original idea of searx and become not better than what searx tries to fix.

**Comment by Cqoicebordel** [comment]
Autocompleter isn't activated by default, so your comment is moot.

**Comment by stef** [comment]
i applaud the sane defaults of searx! yet again, why use a privacy respecting search engine if the only benefit of it you disable?

**Comment by Cqoicebordel** [comment]
Because it's a choice of the user.

**Comment by stef** [comment]
yes, but why does the user want to use the wrong tool? i mean they use google and want privacy, and they use searx and they don't want it, they want js instead. why can't those users who do not care use google, and those who care use searx?

**Comment by Cqoicebordel** [comment]
Because what is wrong for you isn't wrong for them. I know it's hard to hear, but everybody isn't like you.

**Comment by dalf** [comment]
Google, to pick one, is more than a thing that is feed with privacy. 
It's provide useful feature as a counter part.

Now "people" knows that privacy matter but no everyone are willing to pay the price for it.
May be, this is something that should be explain.

About searx : what privacy doesn't it provide ?

In my opinion, searx is proxy to different search engines. So it can 
- avoid tracking cookie / user.
- avoid tracking of the IP.
- avoid tracking of the geographic location.
- avoid ads

What does the auto-completer do : it's a proxy to different auto-completer. Same feature : no cookie, no IP,  no geographic location. 

How does it hurt ? The search engine knows more ? but about who ?

From the user point of view : with the auto-completer there is no need to type everything, only the first letters, no need to remember the right spelling.

Moreover why disabling javascript is good ? because javascript can do evil stuff ? so either 
- you don't trust searx javascript code. Example : someone can modify the searx code to do nasty stuff in javascript.
- you don't trust the "wire". Example : some company / country install proxy with appropriate certificates to intercept https traffic. This kind of appliance can modify the pages on the fly. 

It's all about trust : why do you trust a searx instance ?

Worth case scenario : someone says "Hey look I have setup an searx with bunch of bandwidth and IP", and log everything. You will never know. 

In fact, how searx scale ?
- if everyone has it own instance. I can't find the article, but I remember that Google is able to determine an user with 7 searches, even it is not logged. So which privacy doesn't it provide ?
- at the opposite, if everyone is using the same instance, this unique instance must a bunch of IP to avoid captcha. But in this case, how can we trust this instance ?

Of course, it's not black or white. A searx instance with few users will do the job.

**Comment by stef** [comment]
i guess one way of preserving the privacy of searx users, is by limiting the ways they can shoot themselves in their foot. the not trusting searx, becomes muddled, when you consider other issues, like round-robbin dns searx instances, where you have a bunch of instances being served. would you still trust that js? also lot's of users, have not the privilege to trust their wires. many governments have their CA certs in the browser stores, there are other ways to strip or intercept https even in the enterprise setting, i heard of lenovo and other commercial enterprises doing this as crapware. 

in the case of the autocompleter, there is this other issue what i reflected on earlier, where the request will be routed through searx to preserve privacy, doing that increases latency quite significantly, your auto-completing will be slow.

**Comment by Cqoicebordel** [comment]
I as said earlier, this trade-off is a choice by the user.  
We are not more intelligent than the user, and we are not here to teach him.

We are here to respond to his queries as he wants to. No more, no less.

**Comment by dalf** [comment]
> when you consider other issues, like round-robbin dns searx instances, where you have a bunch of instances being served. would you still trust that js?

Even if there is no javascript in the searx github code, every searx administrator can add javascript, you will never know except if you look at source code of every pages.
In fact, it's not a searx issue, it's browser issue. noscript extension will do the job.

**Comment by stef** [comment]
well, but if you know that searx needs no js by default, then you'll be much happier using it than others where you have to enable noscript. finally search-engine that doesn't need no js, why erode that? i don't think it's the goal of searx to serve even users who want to shoot themselves in the foot, there's enough engines for that already. if you disable the privacy stuff, why use searx in the first place?

why not telling the user how to use searx properly? i mean there's even a popup saying that you have to click after selecting a category, that is very much the same. and i think privacy aspects of searx also mean that searx should not allow or at least warn the user when searx "respond to his queries as he wants to" but violates the promises of searx in the mean time.

**Comment by Cqoicebordel** [comment]
I as said earlier, this trade-off is a choice by the user.

We are not more intelligent than the user, and we are not here to teach him.

We are here to respond to his queries as he wants to. No more, no less.

(Oh wait, déjà vu)

**Comment by stef** [comment]
so we've run out of arguments it seems.

**Comment by pointhi** [comment]
In my opinion, not js cause the privacy issues, programmers who use it to harm privacy are the problem. As before said, even there is no js, no one can fully trust an searx instance which is not controlled by himselfe, because every admin can add own privacy harming code on server-side.

I think, we should always serving basic search-requests without needing js. I have implemented that feature for oscar soon, but I'm also developing new client-side code (only for the oscar-theme), to add new features to searx like a big map, which is also supporting routing and reverse geolocation.

I don't force anyone to use it. If one doesn't like it, he can always switch to themes like default or courgett, or simply disable js, or more simple, not using that new features. (Because without user interaction, no external requests are done).

I think it is a good choise, to have mostly js free themes like default, and themes like oscar which bundle more features, but also require more bandwidth for js-related stuff. In searx, the user choise what he want. In my opition, most potential users are on the side of modern features as for using a js-free site.

Why should we not create a search-site which has similar features like other big search-engines, but without privacy harming? Only because js can be used in an evil way?

**Comment by Glandos** [comment]
What about writing a small warning on this feature? I would see something like:

```
 Warning: Enable this feature will send every keystroke to the remote site. All requests are proxied (i.e. no cookie, no tracking) through the Searx instance, but the remote site could be able to identify you anyway.
```

This is a draft, and not perfect. It should be displayed next to the preference item.

**Comment by ghost** [comment]
I think @Glandos' warning message would be a good idea.

**Comment by Cqoicebordel** [comment]
Agreed, but I would present it differently : if we have to describe in details the risks of each preferences, I'm afraid it would overcharge the preferences' page.

I think a better way of doing it would be to add a FAQ in the about page, and link to it in the preferences with something like this :  
"Note : activating the auto-complete might undermine your privacy. Learn more _here_"

**Comment by pointhi** [comment]
yes, or we add something like a yellow warning-sign to the autocompletion configuration, which is showing that text while hovering.

**Comment by stef** [comment]
https://trisquel.info/en/forum/searx-metasearch-engine#comment-69404

> ... and they managed to make it all js free!

there's other people being excited about not tainting stuff with js.

**Comment by dalf** [comment]
The new version of typeahead.js can handle multiple datasets : [here an example](http://twitter.github.io/typeahead.js/examples/#multiple-datasets)

I'm not sure that two datasets ( = two requests ?) is right to do that, but at least it shown one way to show the responses : one dataset for general result for example, another one for map. 

With this UI, user could choose multiple autocompleters. None by default as now : as said before, it's only for people who would like to use javascript, knowing what's they doing.

About the new version of typeahead.js : 
https://github.com/twitter/typeahead.js/blob/master/CHANGELOG.md#0110-april-25-2015

> There are bunch of API changes with this release so don't expect backwards compatibility with previous versions. 
> ...
> Beware that since this release is pretty much a rewrite, there are bound to be some bugs.

**Comment by stef** [comment]
> it's only for people who would like to use javascript, knowing what's they doing.

statistically speaking the number of users conscious about their js usage is negligible in contrast to the number us users using it out of convenience.

**Comment by dalf** [comment]
@stef your point was understood and taking it into consideration : 
- searx can be use without javascript (look at #171 : one blocking point was javascript, since then oscar was modified to be use without javascript)
- searx doesn't enable autocompleter even if javascript is enabled

The missing point is a warning about privacy and javascript but that's ANOTHER issue, feel free to create one issue about it, or even better to create a warning / documentation about javascript, and create a pull request.
It's not an order, just if you have time it would be great as we are lacking some arms / time. After you can argue on how we spend the time : security, fancy unsecure feature, etc...  but look at the lastest commit : some was about https connection. In my opinion it's far more dangerous to spread requests in clear text from the searx instance than using javascript.

Once again, that's not the issue here. Here it's about how to improve an existing one, even it's for people who want to use javascript (yes the warning / documentation part is still missing, read the previous paragraph, and should we stop thinking about it ? I said thinking).

I saw a lot of question like "which privacy searx provides ?", once again it's another issue. Feel free to help.

This issue is about how to improve autocompleter.

My life:  I wish that this feature could be improved because I use it on my own instance on my own server, and shared it with some friends. BTW, one of my friend has the noscript extension.

EDIT : I feel like your request is to have an searx without javascript, to force every user to use searx without javascript, and that we defend this opinion of rejecting every features that could use javascript. I'm not the only one here, but I think that's not the searx spirit even we have taken the point into consideration (look at the default settings, look at the missing documentation I spoke about).

**Comment by Cqoicebordel** [comment]
Nicely said dalf !

stef, I will shortly propose a change to add more information to the users.

**Comment by roke-julian-lockhart** [comment]
@stef, please stop attempting to prevent addition of this, if your sole reason is privacy. I am unimpacted by any potential lack of privacy, because I merely want to use a good amalgamator of search-results. Currently, SearX performs that task well. It additionally provides privacy for those that desire it, but I do not, so this would not be “shooting [myself] in the foot” as you purport it would.
